asttokens/mark_tokens.py

```diff
@@ -199,15 +202,17 @@ def handle_attr(self, node, first_token, last_token):
   visit_assignattr = handle_attr
   visit_delattr = handle_attr
 
-  def handle_doc(self, node, first_token, last_token):
+  def visit_functiondef(self, node, first_token, last_token):
     # With astroid, nodes that start with a doc-string can have an empty body, in which case we
     # need to adjust the last token to include the doc string.
     if not node.body and getattr(node, 'doc', None):
       last_token = self._code.find_token(last_token, token.STRING)
+    prev = self._code.prev_token(first_token)
```

Comment:
A bit concerned that if `first_token` is the first-ever token, then `prev_token` throws. To guard against it, I suggest wrapping into `if first_token.index > 0:`

Reply:
It doesn't throw, it returns something like `tokens[-1]`. Which is probably a problem that we should fix, just not here. But I can add that check anyway, particularly in case we make it throw in the future.